In [105]:
# Import required libraries and dependencies
import pandas as pd
import numpy as np
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [15]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv("Resources/crypto_market_data.csv",index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [16]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [107]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(width = 800, height = 400, rot = 90)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

In [18]:
#Renaming Column Names
df_market_data = df_market_data.rename(columns={'price_change_percentage_24h': 'change24h', 'price_change_percentage_7d': 'change7d','price_change_percentage_14d': 'change14d', 'price_change_percentage_30d': 'change30d', 'price_change_percentage_60d': 'change60d', 'price_change_percentage_200d': 'change200d', 'price_change_percentage_1y': 'change1y'})


In [23]:
df_market_data.head(25)

change24h  change7d  change14d  change30d  change60d  \
coin_id                                                                  
bitcoin             1.08388   7.60278    6.57509    7.67258   -3.25185   
ethereum            0.22392  10.38134    4.80849    0.13169  -12.88890   
tether             -0.21173   0.04935    0.00640   -0.04237    0.28037   
ripple             -0.37819  -0.60926    2.24984    0.23455  -17.55245   
bitcoin-cash        2.90585  17.09717   14.75334   15.74903  -13.71793   
binancecoin         2.10423  12.85511    6.80688    0.05865   36.33486   
chainlink          -0.23935  20.69459    9.30098  -11.21747  -43.69522   
cardano             0.00322  13.99302    5.55476   10.10553  -22.84776   
litecoin           -0.06341   6.60221    7.28931    1.21662  -17.23960   
bitcoin-cash-sv     0.92530   3.29641   -1.86656    2.88926  -24.87434   
crypto-com-chain    0.61209  -5.67151   -8.53948  -17.44782  -16.47600   
usd-coin           -0.17825  -0.11871   -0.00568   -0.16584    0.04271   
eos                 0.14477  -1.31177    1.13751   -4.63398  -30.16898   
monero              0.42961  15.78515   18.41097   38.95974   41.72500   
tron                0.07647   4.23886    1.40337  -12.60389    5.52545   
tezos              -0.67316   8.95665    2.32062  -14.12663  -44.82248   
okb                -2.72700  -4.55389   -5.02662  -10.43847   -2.83120   
stellar            -1.00843   2.07149   -1.08217   -8.12933  -30.80369   
cosmos             -0.95103  16.08534    5.51074    4.57813   -7.20130   
cdai                0.21169   0.05820    0.17076   -2.18147    0.65726   
neo                 0.49302   2.44243   -9.84803  -21.95472   13.51879   
wrapped-bitcoin     1.10231   7.40537    6.55668    7.37557   -3.58772   
leo-token          -0.13192  -1.34886   -7.02859    3.07525   -7.54455   
huobi-token        -0.40818   1.61798    0.45488   -3.25488   -3.40689   
nem                -0.84990  -0.85140    2.64844   -3.65382   82.86094   

                  change200d   change1y  
coin_id                                  
bitcoin             83.51840   37.51761  
ethereum           186.77418  101.96023  
tether              -0.00542    0.01954  
ripple              39.53888  -16.60193  
bitcoin-cash        21.66042   14.49384  
binancecoin        155.61937   69.69195  
chainlink          403.22917  325.13186  
cardano            264.51418  156.09756  
litecoin            27.49919  -12.66408  
bitcoin-cash-sv      7.42562   93.73082  
crypto-com-chain   226.70782  305.05908  
usd-coin            -0.15691   -0.19205  
eos                 18.06111  -17.56753  
monero             169.52147  141.04116  
tron               132.88436   59.23821  
tezos               43.42842  140.01279  
okb                 39.95853  141.95791  
stellar             84.62157   13.80715  
cosmos             185.99786   82.43833  
cdai                -0.39210   -0.28783  
neo                158.64773  131.29655  
wrapped-bitcoin     83.90520   37.53424  
leo-token           16.40588   21.00263  
huobi-token         42.23704   24.57164  
nem                216.17761  200.71797

### Prepare the Data

In [26]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
normalizeddata = StandardScaler().fit_transform(df_market_data[['change24h', 'change7d','change14d', 'change30d','change60d', 'change200d','change1y']])


In [29]:
# Create a DataFrame with the scaled data
scaleddata = pd.DataFrame(normalizeddata, columns=['change24h', 'change7d','change14d', 'change30d','change60d', 'change200d','change1y'])

# Copy the crypto names from the original data
scaleddata['Crypto Names']=df_market_data.index


# Set the coinid column as index
scaleddata = scaleddata.set_index('Crypto Names')


# Display sample data
scaleddata.head(10)

change24h  change7d  change14d  change30d  change60d  \
Crypto Names                                                            
bitcoin           0.508529  0.493193   0.772200   0.235460  -0.067495   
ethereum          0.185446  0.934445   0.558692  -0.054341  -0.273483   
tether            0.021774 -0.706337  -0.021680  -0.061030   0.008005   
ripple           -0.040764 -0.810928   0.249458  -0.050388  -0.373164   
bitcoin-cash      1.193036  2.000959   1.760610   0.545842  -0.291203   
binancecoin       0.891871  1.327295   0.800214  -0.057148   0.778653   
chainlink         0.011397  2.572251   1.101647  -0.490495  -0.931954   
cardano           0.102530  1.508001   0.648885   0.328959  -0.486349   
litecoin          0.077497  0.334297   0.858520  -0.012646  -0.366477   
bitcoin-cash-sv   0.448952 -0.190684  -0.248043   0.051634  -0.529666   

                 change200d  change1y  
Crypto Names                           
bitcoin           -0.355953 -0.251637  
ethereum          -0.115759 -0.199352  
tether            -0.550247 -0.282061  
ripple            -0.458259 -0.295546  
bitcoin-cash      -0.499848 -0.270317  
binancecoin       -0.188232 -0.225533  
chainlink          0.387759 -0.018284  
cardano            0.065080 -0.155428  
litecoin          -0.486266 -0.292351  
bitcoin-cash-sv   -0.532961 -0.206029

---

### Find the Best Value for k Using the Original Data.

In [30]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [31]:
# Create an empty list to store the inertia values
inertia=[]

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(scaleddata)
    inertia.append(k_model.inertia_)


/Users/JavonSeaforth/opt/anaconda3/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/JavonSeaforth/opt/anaconda3/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/JavonSeaforth/opt/anaconda3/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/JavonSeaforth/opt/anaconda3/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The 

In [33]:
# Create a dictionary with the data to plot the Elbow curve
elbowcurve = {"k_original": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbowcurveddf = pd.DataFrame(elbowcurve)

In [34]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
inertialinechart = elbowcurveddf.hvplot.line(x = "k_original", y = "inertia", title = "Elbow Curve", xticks = k)

inertialinechart

:Curve   [k_original]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer: The best value for k is 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [35]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [36]:
# Fit the K-Means model using the scaled data
model.fit(scaleddata)

/Users/JavonSeaforth/opt/anaconda3/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


KMeans(n_clusters=4, random_state=0)

In [40]:
# Predict the clusters to group the cryptocurrencies using the scaled data
kmeanspredictions = model.predict(scaleddata)

# Print the resulting array of cluster values.
kmeanspredictions

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

In [41]:
# Create a copy of the DataFrame
scaleddataprediction = scaleddata.copy()

In [43]:
# Add a new column to the DataFrame with the predicted clusters
scaleddataprediction['Prediction']=kmeanspredictions


# Display sample data
scaleddataprediction.head()

change24h  change7d  change14d  change30d  change60d  \
Crypto Names                                                         
bitcoin        0.508529  0.493193   0.772200   0.235460  -0.067495   
ethereum       0.185446  0.934445   0.558692  -0.054341  -0.273483   
tether         0.021774 -0.706337  -0.021680  -0.061030   0.008005   
ripple        -0.040764 -0.810928   0.249458  -0.050388  -0.373164   
bitcoin-cash   1.193036  2.000959   1.760610   0.545842  -0.291203   

              change200d  change1y  Prediction  
Crypto Names                                    
bitcoin        -0.355953 -0.251637           2  
ethereum       -0.115759 -0.199352           2  
tether         -0.550247 -0.282061           0  
ripple         -0.458259 -0.295546           0  
bitcoin-cash   -0.499848 -0.270317           2

In [45]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
datascatterplot = scaleddataprediction.hvplot.scatter(x = "change24h", y = "change7d", by = "Prediction", hover_cols = ['Crypto Names'])

datascatterplot

:NdOverlay   [Prediction]
   :Scatter   [change24h]   (change7d,Crypto Names)

---

### Optimize Clusters with Principal Component Analysis.

In [46]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [47]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
marketpca = pca.fit_transform(scaleddata)

# View the first five rows of the DataFrame. 
marketpca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [62]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_
explainedvariance = np.sum(pca.explained_variance_ratio_)
explainedvariancepercentage = explainedvariance * 100
print(f' The Total Explained Variance is:', explainedvariancepercentage)

 The Total Explained Variance is: 89.50316570309842


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 89.50%

In [89]:
# Create a new DataFrame with the PCA data.
pcadf = pd.DataFrame(marketpca, columns = ["PCA1", "PCA2" , "PCA3"])

# Creating a DataFrame with the PCA data
pcadf["Crypto Names"] = df_market_data.index

# Copy the crypto names from the original data
pcadf["Crypto Names"] = df_market_data.index


# Set the coinid column as index
pcadf = pcadf.set_index('Crypto Names')

# Display sample data
pcadf.head(10)

PCA1      PCA2      PCA3
Crypto Names                                 
bitcoin         -0.600667  0.842760  0.461595
ethereum        -0.458261  0.458466  0.952877
tether          -0.433070 -0.168126 -0.641752
ripple          -0.471835 -0.222660 -0.479053
bitcoin-cash    -1.157800  2.041209  1.859715
binancecoin     -0.516534  1.388377  0.804071
chainlink       -0.450711  0.517699  2.846143
cardano         -0.345600  0.729439  1.478013
litecoin        -0.649468  0.432165  0.600303
bitcoin-cash-sv -0.759014 -0.201200 -0.217653

---

### Find the Best Value for k Using the PCA Data

In [86]:
# Create a list with the number of k-values from 1 to 11
kpca = list(range(1,11))


In [83]:
# Create an empty list to store the inertia values
inertia=[]


# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in kpca:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(pcadf)
    inertia.append(k_model.inertia_)


/Users/JavonSeaforth/opt/anaconda3/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/JavonSeaforth/opt/anaconda3/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/JavonSeaforth/opt/anaconda3/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/Users/JavonSeaforth/opt/anaconda3/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The 

In [91]:
# Create a dictionary with the data to plot the Elbow curve
elbowcurve2 = {'k_pca': k_pca , 'inertia': inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbowcurveddf2 = pd.DataFrame(elbowcurve2)
elbowcurveddf2.head(10)


k_pca     inertia
0      1  256.874086
1      2  168.811896
2      3   93.774626
3      4   49.665497
4      5   37.839466
5      6   27.720626
6      7   21.182776
7      8   17.016962
8      9   13.668325
9     10   10.484890

In [93]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
inertialinechart2 = elbowcurveddf2.hvplot.line(x = "k_pca", y = "inertia", title = "Elbow Curve", xticks = k)

inertialinechart2

:Curve   [k_pca]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No it does not

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [94]:
# Initialize the K-Means model using the best value for k
model2 = KMeans(n_clusters=4, random_state=0)


In [96]:
# Fit the K-Means model using the PCA data
model2.fit(pcadf)


/Users/JavonSeaforth/opt/anaconda3/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


KMeans(n_clusters=4, random_state=0)

In [97]:
# Predict the clusters to group the cryptocurrencies using the PCA data
clusterspredict = model2.predict(pcadf)


# Print the resulting array of cluster values.
clusterspredict

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

In [99]:
# Create a copy of the DataFrame with the PCA data
pcdpredictiondf = pcadf.copy()

# Add a new column to the DataFrame with the predicted clusters
pcdpredictiondf['Pca Prediction'] = clusterspredict


# Display sample data
pcdpredictiondf.head(10)


PCA1      PCA2      PCA3  Pca Prediction
Crypto Names                                                 
bitcoin         -0.600667  0.842760  0.461595               2
ethereum        -0.458261  0.458466  0.952877               2
tether          -0.433070 -0.168126 -0.641752               0
ripple          -0.471835 -0.222660 -0.479053               0
bitcoin-cash    -1.157800  2.041209  1.859715               2
binancecoin     -0.516534  1.388377  0.804071               2
chainlink       -0.450711  0.517699  2.846143               2
cardano         -0.345600  0.729439  1.478013               2
litecoin        -0.649468  0.432165  0.600303               2
bitcoin-cash-sv -0.759014 -0.201200 -0.217653               0

In [101]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
pcascatterplot = pcdpredictiondf.hvplot.scatter(x = "PCA1", y = "PCA2", by = "Pca Prediction", hover_cols = ['Crypto Names'])

pcascatterplot

:NdOverlay   [Pca Prediction]
   :Scatter   [PCA1]   (PCA2,Crypto Names)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [102]:
# Composite plot to contrast the Elbow curves
inertialinechart + inertialinechart2


:Layout
   .Curve.I  :Curve   [k_original]   (inertia)
   .Curve.II :Curve   [k_pca]   (inertia)

In [103]:
# Composite plot to contrast the clusters
datascatterplot + pcascatterplot

:Layout
   .NdOverlay.I  :NdOverlay   [Prediction]
      :Scatter   [change24h]   (change7d,Crypto Names)
   .NdOverlay.II :NdOverlay   [Pca Prediction]
      :Scatter   [PCA1]   (PCA2,Crypto Names)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Reducing the features to the cluster had a substantial impact when using the K-Means.  It should be noted that when using PCA the clusters were more clustered together which would give us a more precise and accurate depiction of the data.